File: solutions/typescript/binary-search/1/binary-search.ts
export function find(haystack: number[], needle: number): number | never {
  haystack.sort((a, b) => a - b)
  let left = 0
  let right = haystack.length
  while (left < right) {
    const mid = Math.floor((left + right - 1) / 2)
    if (haystack[mid] === needle) {
      return mid
    }
    if (haystack[mid] < needle) {
      left = mid + 1
    } else {
      right = mid
    }
  }
  throw new Error('Value not in array')
}

File: solutions/python/binary-search/1/binary_search.py
def find(search_list, value):
    left, right = 0, len(search_list)
    while left < right:
        mid = (left + right - 1) >> 1
        match search_list[mid] - value:
            case cmp if cmp > 0:
                right = mid
            case cmp if cmp < 0:
                left = mid + 1
            case cmp if cmp == 0:
                return mid
    raise ValueError("value not in array")